feat: draft pipeline for AWS OS

yevhen-cap opened the pull request:
## Description

<!-- Describe the goal and purpose of this PR. -->

## Checklist

<!--- Please delete options that are not relevant. Boxes should be checked by reviewer. -->
- [ ] for CI changes:
  - [ ] structural/foundational changes signed off by [CI DRI](https://github.com/cmur2)
  - [ ] [ci.yml](https://github.com/camunda/camunda/blob/main/.github/workflows/ci.yml) modifications comply with ["Unified CI" requirements](https://github.com/camunda/camunda/wiki/CI-&-Automation#workflow-inclusion-criteria)

## Related issues

- https://github.com/camunda/team-data-engine-maintenance/issues/49

cmur2 commented on .github/workflows/zeebe-aws-os-dispatch-tests.yml:
```diff
@@ -0,0 +1,109 @@
+name: Zeebe CI (AWS OS)
```
I'd use a different name since the tests this workflow is running are not all related to Zeebe. Maybe just "AWS OpenSearch Dispatch Tests" orso.

cmur2 commented on .github/workflows/zeebe-aws-os-dispatch-tests.yml:
```diff
@@ -0,0 +1,109 @@
+name: Zeebe CI (AWS OS)
+
+on:
+  workflow_dispatch: { }
+  workflow_call: { }
```
This means the workflow is triggered by either humans or another workflow (which is not visible in this PR). It would be useful to know from where you want to trigger it, for a full review.

cmur2 commented on .github/workflows/zeebe-aws-os-dispatch-tests.yml:
```diff
@@ -0,0 +1,109 @@
+    shell: bash
+
+env:
+  DOCKER_PLATFORMS: "linux/amd64,linux/arm64"
```
This variable is unused as far as I can see, please remove when unused.

cmur2 commented on .github/workflows/zeebe-aws-os-dispatch-tests.yml:
```diff
@@ -0,0 +1,109 @@
+            maven-modules: "'io.camunda:zeebe-opensearch-exporter'"
+            maven-build-threads: 2
+            maven-test-fork-count: 7
+            runs-on: ubuntu-latest
+          - group: camunda-exporter
+            name: "Camunda Exporter Integration Tests"
+            maven-modules: "'io.camunda:camunda-exporter'"
+            maven-build-threads: 2
+            maven-test-fork-count: 7
+            runs-on: ubuntu-latest
```
Using `runs-on: ubuntu-latest` for all jobs means that *none of those tests* are running against a real AWS OpenSearch instance - the real AWS OpenSearch instance (available under `CAMUNDA_CI_OPENSEARCH_2_9_URL`) is only reachable from Camunda-hosted runners but not from GitHub.

I'm not sure if it is intentional to not use a real AWS OpenSearch instance?

cmur2 commented on .github/workflows/zeebe-aws-os-dispatch-tests.yml:
```diff
@@ -0,0 +1,109 @@
+          -D maven.javadoc.skip=true
+          -D skipUTs -D skipChecks -D skipOptimize
+          -D failsafe.rerunFailingTestsCount=3 -D flaky.test.reportDir=failsafe-reports
+          -Dcamunda.it.opensearch.aws_instance_url=${{ steps.secrets.outputs.CAMUNDA_CI_OPENSEARCH_2_9_URL }}
```
Apparently this setting doesn't take effect otherwise the job cannot be successful on Github-hosted runners (which don't have access to the real AWS OpenSearch instance provided under `CAMUNDA_CI_OPENSEARCH_2_9_URL`) - but also thir PR doesn't show if the job is successful, because it is not triggered.

cmur2 replied:
The new CI job is failing right now. The error message is `Connect to localhost:443 [localhost/127.0.0.1, localhost/0:0:0:0:0:0:0:1] failed: Connection refused` which to me looks like the test cases assume that the AWS OpenSearch instance is available under `localhost` (which it is not).

I am still wondering if the `camunda.it.opensearch.aws_instance_url` parameter takes any effect. I tried to see if there is any code in the repository that uses/reads this parameter via [GitHub search but there is none](https://github.com/search?q=repo%3Acamunda%2Fcamunda+%22aws_instance_url%22&type=code).

@yevhen-cap where/how did you find that this parameter should be used? Any documentation or another engineer mentioning that it works? I don't see any evidence in the code that this is the right parameter.

yevhen-cap replied:
Merged tests from branch, where this parameter present

cmur2 commented on .github/workflows/zeebe-aws-os-dispatch-tests.yml:
```diff
@@ -0,0 +1,106 @@
+
+on:
+  workflow_dispatch: { }
+  push: { }
```
What is your goal, on which branches should this workflow run automatically? All branches? Just the `main` branch?

If the answer is all branches, then we should use [these triggers](https://github.com/camunda/camunda/blob/main/.github/workflows/ci.yml#L5-L10).
If the answer is just the `main` branch, then limit the existing `on.push` trigger accordingly.

yevhen-cap replied:
At this moment on.push added for pipeline testing. will be removed before merge and on.workflow_dispatch would be only available option

cmur2 commented on .github/workflows/zeebe-aws-os-dispatch-tests.yml:
```diff
@@ -0,0 +1,106 @@
+        id: build-zeebe
+        with:
+          maven-extra-args: -T1C -PskipFrontendBuild -D skipOptimize
+      - uses: ./.github/actions/build-platform-docker
+        with:
+          repository: localhost:5000/camunda/zeebe
+          version: current-test
+          push: true
+          distball: ${{ steps.build-zeebe.outputs.distball }}
```
I'm not sure if this step of building the Docker image is actually needed. We cannot verify it now since the workflow is failing anyways, but let's keep this comment open for later (when the workflow is 🟢): then try the experiment if removing this step still makes this workflow green.

yevhen-cap replied:
This step were removed

igpetrov commented on .github/workflows/zeebe-aws-os-dispatch-tests.yml:
```diff
@@ -0,0 +1,110 @@
+      - name: Maven Test Build
+        run: >
+          url="https://"${{ steps.secrets.outputs.CAMUNDA_CI_OPENSEARCH_2_9_URL }};
+          ./mvnw -B -T ${{ matrix.maven-build-threads }} --no-snapshot-updates
```
For debug trace

igpetrov commented:
Code changes moved from https://github.com/camunda/camunda/pull/25361. The reason: we wanted to make sure both PR and code fit each other, so in this specific case, we decided to union PRs

Commit:
feat: draft pipeline for AWS OS

- removed broken test project
- removed initial pipeline
- change trigger to on.pull_request
- remove build platform docker
- rework OS exporter tests to support AWS

Commit:
fix: removed on.pull_request

cmur2 commented on .github/workflows/zeebe-aws-os-dispatch-tests.yml:
```diff
@@ -0,0 +1,93 @@
+    name: "[IT] ${{ matrix.name }}"
+    timeout-minutes: 15
+    runs-on: ${{ matrix.runs-on }}
+    strategy:
+      fail-fast: false
+      matrix:
+        group: [ os-exporter ]
+        include:
+          - group: os-exporter
+            name: "Zeebe OS Exporter Integration Tests"
+            maven-modules: "'io.camunda:zeebe-opensearch-exporter'"
+            maven-build-threads: 2
+            maven-test-fork-count: 7
+            runs-on: aws-core-8-default
```
(Not a blocker) A matrix build with exactly one group may be refactored to remove the unnecessary matrix setup, unless you plan future additions to the matrix.

igpetrov replied:
Indeed, planned at least 3 more modules, but likely 5

cmur2 commented on .github/workflows/zeebe-aws-os-dispatch-tests.yml:
```diff
@@ -0,0 +1,93 @@
+            runs-on: aws-core-8-default
+    env:
+      ZEEBE_TEST_DOCKER_IMAGE: localhost:5000/camunda/zeebe:current-test
+    services:
+      registry:
+        image: registry:2
+        ports:
+          - 5000:5000
```
This setup (providing a locally available temporary Docker registry) is probably redundant now as well since the Docker image build step got removed.

Commit:
fix: comment resolving

igpetrov reviewed (approved):
LGTM, thank you

---

Let's wait for Camunda engineers approve before merge.

cmur2 reviewed (approved):
Assuming the current [test run](https://github.com/camunda/camunda/actions/runs/12182045073/job/33980102422?pr=24986) passes, the CI code looks also good to me!

igpetrov commented:
Merging for now to unblock test development of another modules. If there are issues, happy to address them in another PR.

cc: @ChrisKujawa

ChrisKujawa reviewed (commented):
Better late then never I guess :) LGTM. Just two comments (for potential improvement)

igpetrov commented on zeebe/exporters/opensearch-exporter/src/test/java/io/camunda/zeebe/exporter/opensearch/SearchDBExtension.java:
```diff
@@ -0,0 +1,52 @@
+    implements BeforeAllCallback, BeforeEachCallback, AfterEachCallback {
+
+  protected static final String IT_OPENSEARCH_AWS_INSTANCE_URL_PROPERTY =
+      "camunda.it.opensearch.aws_instance_url";
```
Addressed here: https://github.com/camunda/camunda/pull/25802

igpetrov commented on zeebe/exporters/opensearch-exporter/src/test/java/io/camunda/zeebe/exporter/opensearch/OpensearchClientIT.java:
```diff
@@ -131,45 +92,50 @@ void shouldPutIndexTemplate() {
     assertIndexTemplate(template, expectedTemplate);
   }
 
   @Test
   void shouldAuthenticateWithBasicAuth() {
+    // we currently only allow role-based authentication in our AWS OS ITs
+    if (searchDB.isAwsTest()) {
+      return;
+    }
```
Addressed here: https://github.com/camunda/camunda/pull/25802
